src/typing/Typing.ml

```diff
@@ -22,22 +22,19 @@ module G = AST_generic
 let rec type_of_expr lang e : G.name Type.t * G.ident option =
   match e.G.e with
   | G.L lit ->
-      let t =
-        match lit with
-        (* NB: We could infer Type.Number for JS int/float literals, but we can
-         * handle that relationship in matching and we can be more precise for
-         * now. One actual rule uses `float` for a typed metavariable in JS so
-         * let's avoid breaking that for now at least. *)
-        | G.Int _ -> Type.Builtin Type.Int
-        | G.Float _ -> Type.Builtin Type.Float
-        | G.Bool _ -> Type.Builtin Type.Bool
-        | G.String _ -> Type.Builtin Type.String
-        | _else_ -> Type.NoType
-      in
+      let t = type_of_lit lit in
       (t, None)
+  | G.DotAccess
+      (_obj, _, FN (Id (("length", _), { id_type = { contents = None }; _ })))
+    when lang =*= Lang.Java ->
+      (* TODO: Fix guess_type_of_dotaccess to take an "expected type" (as in bidirectional
+       * type checking) so that we can distinguish `length` as a method (e.g. `String`) and
+       * `length` as an attribute (arrays). *)
+      (Type.Builtin Type.Int, None)
```

Comment:
I'd really like to keep the guessing parts of type inference separate. This is a heuristic, not a part of proper type inference. Plus, won't this make it so that we never guess `String.length` to be a method?

Is there a reason that this can't go in `guess_type_of_dotaccess` and instead of using bidirectional type inference, we check the type of the receiver and use that to inform the guess?